tests/test_traverse_project.py

```diff
@@ -510,6 +512,66 @@ class TraverseProjectVector:
             ".venvs/another-venv/lib/python3.11/site-packages",
         },
     ),
+    #
+    # Test invalid 'ignores':
+    #
+    TraverseProjectVector(
+        "empty_ignore_pattern__raises_IgnoreRuleMissing",
+        "empty",
+        code=set(),
+        deps=set(),
+        pyenvs=set(),
+        ignores={"", "\t", "    "},
+        expect_raised=IgnoreRuleMissing,
+    ),
+    TraverseProjectVector(
+        "comment_ignore_pattern__raises_IgnoreRuleMissing",
+        "empty",
+        code=set(),
+        deps=set(),
+        pyenvs=set(),
+        ignores={"# a comment", "# another comment"},
+        expect_raised=IgnoreRuleMissing,
+    ),
+    TraverseProjectVector(
+        "anchored_ignore_pattern_without_basedir__raises_IgnoreRuleError",
+        "empty",
+        code=set(),
+        deps=set(),
+        pyenvs=set(),
+        ignores={"foo/bar"},
```

Comment:
How would this not raise the `IgnoreRuleError`?

Reply:
In order to make `"foo/bar"` a valid pattern in this context, we would have to specify a `base_dir` (required when the pattern is "anchored", i.e. when it contains a slash). You can see tests of this logic in `test_dir_traversal.py`, specifically look at `anchored_pattern__must_have_base_dir` and `ignore_pattern_without_slash__does_not_match_above_base_dir` for examples without/with `base_dir` respectively.

But this logic is not _yet_ available to users of `traverse_project`. At this point in the series of PRs, we can only specify ignore patterns in the CLI, and from the CLI I believe it is _unclear_ what the correct `base_dir` should be: The natural choice would be the current dir, but as soon as you pass multiple `basepath`s on the command line, and/or multiple `--code` or `--deps` paths, I think it becomes impossible to determine which `base_dir` to be used for `--ignore foo/bar`. Therefore, I've chosen to leave `base_dir` unspecified in this case, with the consequence that anchored patterns cannot be given directly to `--ignore` on the command-line.

This is reflected in our code as of this PR: `DirectoryTraversal.ignore()` does take an optional `base_dir`, but this argument is not yet used in `traverse_project.find_sources()`, and it will only start to be used when I add the logic for parsing `.gitignore` files (in the final `--ignore-from` PR that I'm still working on). When parsing `.gitignore` files, the `base_dir` is indeed well-specified: IIRC, [the `gitignore` documentation](https://git-scm.com/docs/gitignore) states that patterns in a file use the parent directory of that file as their `base_dir`.